multicluster/controllers/multicluster/member/serviceexport_controller.go

```diff
@@ -256,46 +269,50 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	// We also watch Service events via events mapping function.
 	// Need to check cache and compare with cache if there is any change for Service.
-	var svcNoChange, epNoChange bool
+	var skipUpdateSvcResourceExport, skipUpdateEPResourceExport bool
 	svcExportNSName := common.NamespacedName(r.leaderNamespace, svcResExportName)
 	epExportNSName := common.NamespacedName(r.leaderNamespace, epResExportName)
 	if svcInstalled {
 		installedSvc := svcObj.(*svcInfo)
-		if apiequality.Semantic.DeepEqual(svc.Spec.Ports, installedSvc.ports) &&
-			apiequality.Semantic.DeepEqual(svc.Spec.ClusterIPs, installedSvc.clusterIPs) {
+		if apiequality.Semantic.DeepEqual(svc.Spec.Ports, installedSvc.ports) {
 			klog.InfoS("Service has been converted into ResourceExport and no change, skip it", "service",
 				req.String(), "resourceexport", svcExportNSName)
-			svcNoChange = true
+			skipUpdateSvcResourceExport = true
 		}
 	}
 
-	if r.endpointIPType == "ClusterIP" {
-		svcIPAsSubset := common.GetServiceEndpointSubset(svc)
+	var newSubsets []corev1.EndpointSubset
+	if r.endpointIPType == common.EndpointIPTypeClusterIP {
+		svcIPAsSubset := getClusterIPEndpointSubset(svc)
 		if len(svcIPAsSubset.Addresses) == 0 {
 			err = r.updateSvcExportStatus(ctx, req, serviceNoClusterIP)
 			if err != nil {
 				return ctrl.Result{}, err
 			}
 			return ctrl.Result{}, nil
 		} else {
-			eps.Subsets = []corev1.EndpointSubset{svcIPAsSubset}
+			newSubsets = []corev1.EndpointSubset{svcIPAsSubset}
 		}
 	} else {
-		newSubsets := common.FilterEndpointSubsets(eps.Subsets)
-		if epsInstalled {
-			installedEp := epsObj.(*epInfo)
-			if apiequality.Semantic.DeepEqual(newSubsets, installedEp.subsets) {
-				klog.InfoS("Endpoints has been converted into ResourceExport and no change, skip it", "endpoints",
-					req.String(), "resourceexport", epExportNSName)
-				epNoChange = true
-			}
+		if r.endpointSliceEnabled {
+			newSubsets = eps.Subsets
+		} else {
+			newSubsets = filterEndpointSubsets(eps.Subsets)
```

Comment:
Can we process Endpoints in one place and stick to use `newSubsets` to cache the new subsets? The current logic is distributed at L242-L268 and here, and somestimes use `eps.Subsets` to cache the result but sometimes set `newSubsets` to `eps.Subsets`.

Reply:
Done